File: frontend-reshet/src/services/credentials.ts
import { httpClient } from "./http";
import type { ControlPlaneListResponse, ControlPlaneListView } from "./types";

export type IntegrationCredentialCategory = "llm_provider" | "vector_store" | "tool_provider" | "custom";

export interface IntegrationCredential {
  id: string;
  organization_id?: string | null;
  category: IntegrationCredentialCategory;
  provider_key: string;
  provider_variant?: string | null;
  display_name: string;
  credential_keys: string[];
  is_enabled: boolean;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateCredentialRequest {
  category: IntegrationCredentialCategory;
  provider_key: string;
  provider_variant?: string | null;
  display_name: string;
  credentials: Record<string, unknown>;
  is_enabled?: boolean;
  is_default?: boolean;
}

export interface UpdateCredentialRequest {
  category?: IntegrationCredentialCategory;
  provider_key?: string;
  provider_variant?: string | null;
  display_name?: string;
  credentials?: Record<string, unknown>;
  is_enabled?: boolean;
  is_default?: boolean;
}

export interface CredentialStatus {
  id: string;
  category: IntegrationCredentialCategory;
  provider_key: string;
  provider_variant?: string | null;
  is_enabled: boolean;
  is_default: boolean;
  updated_at: string;
}

export interface CredentialUsageModelProvider {
  binding_id: string;
  model_id: string;
  model_name: string;
  provider: string;
  provider_model_id: string;
}

export interface CredentialUsageKnowledgeStore {
  store_id: string;
  store_name: string;
  backend: string;
}

export interface CredentialUsageTool {
  tool_id: string;
  tool_name: string;
  builtin_key?: string | null;
  implementation_type?: string | null;
}

export interface CredentialUsageResponse {
  credential_id: string;
  model_providers: CredentialUsageModelProvider[];
  knowledge_stores: CredentialUsageKnowledgeStore[];
  tools: CredentialUsageTool[];
}

export const credentialsService = {
  async listCredentials(
    category?: IntegrationCredentialCategory,
    params?: { skip?: number; limit?: number; view?: ControlPlaneListView }
  ): Promise<ControlPlaneListResponse<IntegrationCredential>> {
    const query = new URLSearchParams();
    if (category) query.set("category", category);
    query.set("skip", String(params?.skip ?? 0));
    query.set("limit", String(params?.limit ?? 20));
    query.set("view", params?.view ?? "summary");
    const queryString = query.toString();
    const path = `/admin/settings/credentials${queryString ? `?${queryString}` : ""}`;
    return httpClient.get<ControlPlaneListResponse<IntegrationCredential>>(path);
  },

  async createCredential(data: CreateCredentialRequest): Promise<IntegrationCredential> {
    return httpClient.post<IntegrationCredential>("/admin/settings/credentials", data);
  },

  async updateCredential(id: string, data: UpdateCredentialRequest): Promise<IntegrationCredential> {
    return httpClient.patch<IntegrationCredential>(`/admin/settings/credentials/${id}`, data);
  },

  async getCredentialUsage(id: string): Promise<CredentialUsageResponse> {
    return httpClient.get<CredentialUsageResponse>(`/admin/settings/credentials/${id}/usage`);
  },

  async deleteCredential(id: string, opts?: { force_disconnect?: boolean }): Promise<void> {
    const query = new URLSearchParams();
    if (opts?.force_disconnect) query.set("force_disconnect", "true");
    const suffix = query.toString() ? `?${query.toString()}` : "";
    await httpClient.delete(`/admin/settings/credentials/${id}${suffix}`);
  },

  async listCredentialStatus(): Promise<CredentialStatus[]> {
    return httpClient.get<CredentialStatus[]>("/admin/settings/credentials/status");
  },
};

File: backend/talmudpedia_control_sdk/credentials.py
from __future__ import annotations

from typing import Any, Dict, Optional

from .types import RequestOptions, ResponseEnvelope


class CredentialsAPI:
    def __init__(self, client: Any) -> None:
        self._client = client

    def list(
        self,
        category: Optional[str] = None,
        *,
        skip: int = 0,
        limit: int = 20,
        view: str = "summary",
    ) -> ResponseEnvelope:
        params: Dict[str, Any] = {"skip": skip, "limit": limit, "view": view}
        if category:
            params["category"] = category
        return self._client.request("GET", "/admin/settings/credentials", params=params)

    def create(self, spec: Dict[str, Any], options: Optional[RequestOptions] = None) -> ResponseEnvelope:
        return self._client.request(
            "POST",
            "/admin/settings/credentials",
            json_body=spec,
            options=options,
            mutation=True,
        )

    def update(
        self,
        credential_id: str,
        patch: Dict[str, Any],
        options: Optional[RequestOptions] = None,
    ) -> ResponseEnvelope:
        return self._client.request(
            "PATCH",
            f"/admin/settings/credentials/{credential_id}",
            json_body=patch,
            options=options,
            mutation=True,
        )

    def delete(
        self,
        credential_id: str,
        *,
        force_disconnect: bool = False,
        options: Optional[RequestOptions] = None,
    ) -> ResponseEnvelope:
        return self._client.request(
            "DELETE",
            f"/admin/settings/credentials/{credential_id}",
            params={"force_disconnect": force_disconnect},
            options=options,
            mutation=True,
        )

    def usage(self, credential_id: str) -> ResponseEnvelope:
        return self._client.request("GET", f"/admin/settings/credentials/{credential_id}/usage")

    def status(self) -> ResponseEnvelope:
        return self._client.request("GET", "/admin/settings/credentials/status")
